README.md
# Causal Loop Diagram with roughjs

rough_cld / You can try this [demo](https://assets.eloqueasy.de/rough_cld/).

You can draw causal loop diagrams (CLD). This library uses roughjs for handdrawn circles and connections. You can draw multiple
diagrams on one page. Rough_cld is based on Bootstrap (>= 5.2).

* You can show the diagram completely expanded. For every node and edge you can add some informations, which are displayed as popovers.
* Or you can show the diagram stepwise. In this case informations to each node or edge are shown beside the diagram. If you want you can go backward.

![Example for causal loop](screenshot.png "Causal loop diagram in rough mode")




# Use

The head of your HTML page should look like this:

    <head>
      <title>SVG Test</title>
      <meta charset="utf-8">

      <link href="https://fonts.eloqueasy.de/font-architects-daughter.css" rel="stylesheet">
      <link rel="stylesheet" href="https://assets.eloqueasy.de/bootstrap-5/dist/css/bootstrap.min.css">
      <link href="https://assets.eloqueasy.de/fontawesome-free/css/all.min.css" rel="stylesheet"/>
      <script src="https://assets.eloqueasy.de/roughjs/bundled/rough.js"></script>
      <script src="rough_cld.js"></script>
      <script src="https://assets.eloqueasy.de//bootstrap-5/dist/js/bootstrap.bundle.min.js"></script>
      <script>
          var popoverTriggerList = [].slice.call(document.querySelectorAll('[data-bs-toggle="popover"]'))
          var popoverList = popoverTriggerList.map(function (popoverTriggerEl) {
              return new bootstrap.Popover(popoverTriggerEl)
          })
      </script>
      <style>
          a:link {
            text-decoration: none;
          }
      </style>
    </head>

First you have to declare a canvas and info area. The names must be unique on this page (*svg_456*), and the info div has to be named in the following way: <name of the svg canvas>_info (*svg_456_info*):

    <div class="row mt-3">
        <div class="col-md-9">
            <svg id="svg_456" width="600px" height="400px" viewbox="100 200 600 400"></svg>
        </div>
        <div class="col-md-3">
            <div id="svg_456_info"></div>
        </div>
    </div>

Next you have to include the data:

    <script>
        var rcld_456 = new RoughCld("svg_456")

        nodes = [
            [0, 500, 400, '#c49a4c', "#", "Hasen", "Hasen sind wichtig für das Überleben der Füchse."],
            [1, 200, 400, '#634a1d', "#93753b", "Füchse", "Füchse ernähren sich z.B. von Hasen."]
        ]
        edges = [
            [0, 1, -100, 1, '#', 'Gibt es genug Hasen, so können Füchse ebenfalls leben.'],
            [1, 0, 100, -1, '#', 'Die Füchse verhindern, dass sich die Hasen zu stark vermehren.']
        ]
        rcld_456.draw_cld(nodes, edges, true)
    </script>

With
  
    var rcld_456 = new RoughCld("svg_456")

you have to use the name of your canvas div (*svg_456*). If you want to draw more then one causal loop on one page you have to declare an unique name for each class instance (*rcld_456*).
  
Nodes have the following structure:
  
  - First item is an id.
  - Second you describe the x position.
  - Then you declare the y position.
  - Every node can have a different color.
  - Next you describe the URL for more informations for this node. Normally you fill in "#".
  - Then you declare the title of your popver dialog.
  - Last you write the additional infos for this node.
  
Edges have a similar structure:
  
  - First item is an id from which node the connection is starting.
  - Second id describes the target of your connection.
  - With the *bend* parameter you tell the connector how much it is curved. With "+" and "-" you can change the side of the curved connection (100 vs -100).
  - "1" means strengthening influence (+); -"1" means weakening influence. If you want a bolder curve use "2" or "-2".
  - Next you describe the URL for more informations for this edge. Normally you fill in "#".
  - Then you declare the title of your popver dialog.
  - Last you write the additional infos for this node.

If you want to show the complete diagram (*true*), you call
  
    rcld_456.draw_cld(nodes, edges, true)

If you want to show the diagram stepwise you have to create control buttons. Since index.html has two diagrams on the same page, I had to declare a second class (*rcld_123*):
  
    <div class="col-12">
        <div class="btn-group" role="group" aria-label="forward / backward">
            <button class="btn btn-light" onclick="rcld_123.prev_edge()"><i class="fas fa-chevron-circle-left"></i></button>
            <button class="btn btn-light" onclick="rcld_123.next_edge()"><i class="fas fa-chevron-circle-right"></i></button>
        </div>
    </div>  
  
The last step is calling draw_cld with the complete parameter set to *false*:
  
    rcld_123.draw_cld(nodes, edges, false)
  

# Python

You can generate the nodes and edges by an additional Python program: *python generate_cld.py* . The model has to be described in the following way:

    R = 60
    causal_raw = """
    Regen (1;1;#;Der Regen ist die Basis allen Lebens)
    Weide (3;2;#;Nahrungsgrundlage für viele Beutetiere)
    Beute (4;4;#;Nahrungsgrundlage für viele Jäger)
    Jäger (1;4;#;Spitze der Nahrungspyramide;#a00000)
    Regen +> Weide (2;#;Der Regen fördert das Wachstum der Pflanzen)
    Weide +> Beute (2;#;Eine üppige Wiese kann viele Beutetiere ernähren)
    Beute ++> Jäger (6;#;Viele Beutetiere sind eine gute Nahrungsgrundlage für Jäger)
    Jäger --> Beute (6;#;Jäger hält die Anazhl der Beutetiere im Gleichgewicht)
    """

* The coordinates are calculated depending on *R*.
* You don't have to use apostrophes for strings.
* The *strength* is expressed by *++>* or *-->*. You shuoldn't use more then 3 plus or minus signs.

# See also

* [Roughjs](https://roughjs.com/)
* [Loopy](https://ncase.me/loopy/)
* [Fork of Loopy](https://github.com/schucan/loopy)


CHANGES.txt
rough_cld_steps.js

0.89 2023-09-01 Text for first node and edge
0.88 2022-09-21 URL_SNS inside class
0.87 2022-09-04 "javascript://" instead of "#" for popovers
                https://stackoverflow.com/questions/12590523/clicking-a-popover-scrolls-back-to-top-of-the-page-bootstrap-and-django

0.86 2022-09-02 generate_cld.py uses input and output files
                modified input structure for generate_cld.py
                
0.85 2022-09-01 <strength> may be in the range from -3 .. -1 to 1 .. 3
                correct math errors

0.80 2022-08-30 Published this version


rough_cld.js
/*
 * rough_cld_steps.js
 * 
 * Author: Hans Rauch
 * License: MIT License
 * Created: 2022-08-30
 * Last modified: 2022-09-21
 * Version: 0.88
 *
 * Draws causal loop diagrams in rough mode.
*/
 

class RoughCld {
    
    URL_SNS = "http://www.w3.org/2000/svg"
    
    info = null
    svg = null      // canvas
    rc = null       // using the roughjs library
    popover = null  // popovers by bootstrap
    // styling
    font_family = "'Architects Daughter', sans-serif"
    plus_color = '#bfee3f'
    minus_color = '#f562a3'
    strength_fill_style = 'cross-hatch'

    r = 60        // radius for nodes
    nodes = []
    edges = []

    constructor(canvas_name) {
        this.info = document.getElementById(canvas_name + "_info")
        this.svg = document.getElementById(canvas_name)
        this.rc = rough.svg(this.svg)
    }

    get_popover(url, title, info) {
        // create bootstrap popover
        let link = document.createElementNS(this.URL_SNS, "a");
        link.setAttributeNS('http://www.w3.org/1999/xlink', 'xlink:href', url);
        link.setAttributeNS(null,"tabindex", "0")
        link.setAttributeNS(null,"class", "d-inline-block")
        link.setAttributeNS(null,"data-bs-custom-class", "custom-popover")
        link.setAttributeNS(null,"data-bs-placement", "top")
        link.setAttributeNS(null,"data-bs-toggle", "popover")
        link.setAttributeNS(null,"data-bs-trigger", "focus")
        link.setAttributeNS(null,"data-bs-title", title)
        link.setAttributeNS(null,"data-bs-content", info)
        this.popover = new bootstrap.Popover(link)
        return this.svg.appendChild(link)
    }

    create_node(node) {
        // create and draw node for causal diagram
        let link = null
        let svg_node = null
        
        if ((node.url != "" ) || (node.info != "")) {
            link = this.get_popover(node.url, node.title, node.info)
        } else {
            link = this.svg
        }

        // white background for popover-link
        svg_node = document.createElementNS(this.URL_SNS, 'circle')
        svg_node.setAttributeNS(null,"cx", node.x)     
        svg_node.setAttributeNS(null,"cy", node.y) 
        svg_node.setAttributeNS(null,"r", this.r) 
        svg_node.setAttributeNS(null, "fill", "white")
        link.appendChild(svg_node)
        
        // rough circle
        svg_node = this.rc.circle(node.x, node.y, 2 * this.r, {fill: node.color})
        link.appendChild(svg_node)
        
        // title
        svg_node = document.createElementNS(this.URL_SNS, 'text')
        svg_node.setAttributeNS(null,"x", node.x)     
        svg_node.setAttributeNS(null,"y", node.y) 
        svg_node.setAttributeNS(null,"dy", 24 / 4) 
        svg_node.setAttributeNS(null, "font-size", "24")
        svg_node.setAttributeNS(null, "font-family", this.font_family)
        svg_node.setAttributeNS(null, "text-anchor", "middle")
        svg_node.innerHTML = node.title
        link.appendChild(svg_node)
        return link
    }

    create_edge (edge) {
        // create and draw connection desscribed by edge
        let link = null
        let node = null
        let Cb = {}

        let from = edge.from
        let to = edge.to
        // some maths
        let A = {x: this.nodes[from].x, y: this.nodes[from].y}
        let C = {x: this.nodes[to].x, y: this.nodes[to].y}
        let dx = C.x - A.x
        let dy = C.y - A.y
        let w = Math.sqrt(dx*dx + dy*dy)
        let h = A.y - C.y
        // Base point for center perpendicular
        let Bac = {x: A.x + dx /2, y: A.y + dy / 2}
        // B is in the middle of A and C and is <bend> away from the connecting line
        let v = 0
        if (w != 0) {
            v = Math.asin(h / w)
            let v2 = Math.atan(dy / dx)
        }
        if (Bac.x <= A.x) {
            v += Math.PI / 2
        } else {
            if (Bac.y <= C.y) {
                if (Bac.y == C.y) {
                    v -= Math.PI / 2
                }
            } else {
                v += Math.PI
            }
        }

        let B = {x: 0, y: 0}
        
        let b_slope = 0
        if (dx != 0) {
            if (dy == 0) {
                if (dx > 0) {
                    b_slope = Math.PI / 2
                } else {
                    b_slope = -Math.PI / 2
                }
            } else {
                b_slope = -1 / (dy / dx)
            }
        } else {
            b_slope = 0
        }
        if (dy > 0) {
            B.x = Bac.x + edge.bend * Math.cos(b_slope)
            B.y = Bac.y + edge.bend * Math.sin(b_slope)
        } else {
            B.x = Bac.x - edge.bend * Math.cos(b_slope)
            B.y = Bac.y - edge.bend * Math.sin(b_slope)
        }
        // Angle between points C and B
        let vCB = Math.atan((C.y - B.y) / (C.x - B.x))
        // correct the angle
        if ((C.x > A.x) || ((C.x == A.x) && (C.y < A.y))) {
            vCB += Math.PI
        }
        // Cb is on the circle (this.r+5)
        if (edge.bend < 0) {
            vCB += Math.PI
            if (C.x == A.x){
                vCB -= Math.PI
            }
            Cb = {x: C.x - (this.r + 5) * Math.cos(vCB), y: C.y - (this.r + 5) * Math.sin(vCB)}
        } else {
            Cb = {x: C.x + (this.r + 5) * Math.cos(vCB), y: C.y + (this.r + 5) * Math.sin(vCB)}
        }

        // draw connecting line
        if (edge.url != "") {
            link = this.get_popover(edge.url, this.nodes[edge.from].title + " - " + this.nodes[edge.to].title , edge.info)
        } else {
            link = this.svg
        }

        // curved connecting line
        let points = [[A.x, A.y], [B.x, B.y], [Cb.x, Cb.y]]
        let b = this.rc.curve(points, {stroke: 'black', strokeWidth: Math.abs(edge.strength), roughness: 1.5})
        link.appendChild(b)

        // Arrow
        let a_left = 0
        let a_right = 0
        let aLength = 20
        let l = null
        if (edge.bend > 0) {
            a_left = vCB - 0.3 + Math.PI
            a_right = vCB + 0.6 + Math.PI
            l = this.rc.line(Cb.x, Cb.y, Cb.x - aLength * Math.cos(a_left), Cb.y - 20 * Math.sin(a_left), {strokeWidth: 2})
            b.appendChild(l)
            l = this.rc.line(Cb.x, Cb.y, Cb.x - aLength * Math.cos(a_right), Cb.y - 20 * Math.sin(a_right), {strokeWidth: 2})
            b.appendChild(l)
        } else {
            a_left = vCB - 0.6 + Math.PI
            a_right = vCB + 0.3 + Math.PI
            l = this.rc.line(Cb.x, Cb.y, Cb.x + aLength * Math.cos(a_left), Cb.y + 20 * Math.sin(a_left), {strokeWidth: 2})
            b.appendChild(l)
            l = this.rc.line(Cb.x, Cb.y, Cb.x + aLength * Math.cos(a_right), Cb.y + 20 * Math.sin(a_right), {strokeWidth: 2})
            b.appendChild(l)
        }
        // while developing and testing
        //l = this.rc.line(Bac.x, Bac.y, B.x, B.y, {strokeWidth: 1, stroke: 'red'})
        //b.appendChild(l)
        //l = this.rc.circle(B.x, B.y, 20, {strokeWidth: 1, stroke: 'purple'})
        //b.appendChild(l)

        // white background for popover-link
        node = document.createElementNS(this.URL_SNS, 'circle')
        node.setAttributeNS(null,"cx", B.x)     
        node.setAttributeNS(null,"cy", B.y) 
        node.setAttributeNS(null,"r", 20) 
        node.setAttributeNS(null, "fill", "white")
        link.appendChild(node);        
        // rough circle
        if (edge.strength > 0) {
            node = this.rc.circle(B.x, B.y, this.r-20, {fill: this.plus_color, fillStyle: this.strength_fill_style})
        } else {
            node = this.rc.circle(B.x, B.y, this.r - 20, {fill: this.minus_color, fillStyle: this.strength_fill_style})
        }
        link.appendChild(node)
        // + or -
        node = document.createElementNS(this.URL_SNS, 'text')
        node.setAttributeNS(null,"x", B.x)
        node.setAttributeNS(null,"y", B.y) 
        node.setAttributeNS(null,"dy", 28 / 4) 
        node.setAttributeNS(null, "font-size", "28")
        node.setAttributeNS(null, "font-family", this.font_family)
        node.setAttributeNS(null, "text-anchor", "middle")
        if (edge.strength > 0) {
            node.innerHTML = "+"
        } else {
            node.innerHTML = "-"
        }
        link.appendChild(node)
        return link
    }

    draw_cld(l_nodes, l_edges, complete) {
        // draw causal loop diagram
        let from = 0
        let to = 0
        let t = []
        let node = {}
        this.nodes = []
        for (let i=0; i < l_nodes.length; i++) {
            t = l_nodes[i]
            node = {id: t[0], x: t[1], y: t[2], color: t[3], url: t[4], title: t[5], info: t[6]}
            this.nodes.push(node)
        }
        let edge = {}
        this.edges = []
        for (let i=0; i < l_edges.length; i++) {
            t = l_edges[i]
            edge = {from: t[0], to: t[1], bend: t[2], strength: t[3], url: t[4], info: t[5]}
            this.edges.push(edge)
        }
        if (complete) {
            for (let i=0; i < this.edges.length; i++) {
                this.create_edge(this.edges[i])
            }
            for (let i=0; i < this.nodes.length; i++) {
                this.create_node(this.nodes[i])
            }
        } else {
            this.next_edge()
        }
    }

    // draw causal loop stepwise

    n_curr = 0
    obj_edges = []
    obj_nodes = []

    next_edge() {
        // draw causal loop stepwise forward
        let from = 0
        let to = 0
        let text = ""
        let curr = Math.floor(this.n_curr / 2)
        if (curr < this.edges.length) {
                this.obj_edges.push(this.create_edge(this.edges[curr]))
                this.obj_nodes.push(this.create_node(this.nodes[from]))
                text += "<h4>" + this.nodes[from].title + " &rarr; " + this.nodes[to].title + "</h4>\n"
                if (this.n_curr == 0) {
                    text += "<p>" + this.nodes[from].info + " &rarr; " + this.edges[curr].info + "</p>"
                } else {
                    text += "<p>" + this.edges[curr].info + "</p>\n"
                }
            } else {
                this.obj_nodes.push(this.create_node(this.nodes[to]))
                text += "<h3>" + this.nodes[to].title + "</h3>\n"
                text += "<p>" + this.nodes[to].info + "</p>"
            }
            this.info.innerHTML = text
            this.n_curr += 1
        }
    }

    prev_edge() {
        // draw causal loop stepwise backward    
        let from = 0
        let to = 0
        let obj = null
        let text = ""
        if (this.n_curr > 1) {
            this.n_curr -= 1
            let curr = Math.floor(this.n_curr / 2)
            from = this.edges[curr].from
            to = this.edges[curr].to
            if (this.n_curr % 2 == 0) {
                obj = this.obj_edges.pop()
                this.svg.removeChild(obj)
                obj = this.obj_nodes.pop()
                this.svg.removeChild(obj)
                if (this.n_curr > 0) {
                    text += "<h3>" + this.nodes[from].title + "</h3>\n"
                    text += "<p>" + this.nodes[from].info + "</p>"
                }
            } else {
                obj = this.obj_nodes.pop()
                this.svg.removeChild(obj)
                text += "<h4>" + this.nodes[from].title + " &rarr; " + this.nodes[to].title + "</h4>\n"
                text += "<p>" + this.edges[curr].info + "</p>\n"
            }
            this.info.innerHTML = text
        }
    }
}


generate_cld.py
"""
generate_cld.py

Generates causal loop diagramm in HTML format based on input file

Hans Rauch
(c) MIT LICENSE
2022-09-02
Version 0.86
"""

import random
import re
import sys

causal_raw = """
R = 60
Regen (1;1;Der Regen ist die Basis allen Lebens)
Weide (3;2;Nahrungsgrundlage für viele Beutetiere)
Beute (4;4;Nahrungsgrundlage für viele Jäger)
Jäger (1;4;Spitze der Nahrungspyramide;#a00000)

Regen +> Weide (Der Regen fördert das Wachstum der Pflanzen;2)
Weide +> Beute (Eine üppige Wiese kann viele Beutetiere ernähren;2)
Beute ++> Jäger (Viele Beutetiere sind eine gute Nahrungsgrundlage für Jäger)
Jäger --> Beute (Jäger hält die Anzahl der Beutetiere im Gleichgewicht)
"""

def get_node_id(nodes: list, name: str) -> int:
    """
    find the node-id for <name>
    """
    for n in nodes:
        if n[5] == name:
            return n[0]
    return -1


def generate_causal_diagram(raw: str) -> str:
    """
    generate nodes and edges
    """
    nodes = []
    edges = []
    UID = 'r' + str(random.random())[2:11]
    base_color = '#e0e0ff'

    R = 60
    lines = raw.strip().split('\n')
    for line in lines:
        line = line.strip()
        # read radius
        n_pos = line.find('=')
        if n_pos > 0 and line.startswith('R'):
            line = line[n_pos+1:]
            try:
                R = int(line.strip())
            except:
                pass
            line = ""
        if line:
            com = re.search(r'[-,+]+>+', line)
            if com:
                bend = 50
                url = ''
                info = ''
                name_from = line[:com.start()].strip().strip()
                name_from_id = get_node_id(nodes, name_from)
                strength = len(line[com.start():com.end()].replace('>', ''))
                if line[com.start()] == '+':
                    strength = 1 * strength
                else:
                    strength = -1 * strength
                line = line[com.end():].strip()
                com = re.search(r'[\( ]', line)
                if not com:
                    name_to = line.strip()
                    name_to_id = get_node_id(nodes, name_to)
                else:
                    n_start = com.start()
                    name_to = line[:n_start].strip()
                    name_to_id = get_node_id(nodes, name_to)
                    line = line[n_start:].strip()
                    com = re.search(r'\(.*\)', line)
                    if com:
                        line = line[1:-1]
                        items = line.split(";")
                        info = items[0].strip()
                        try:
                            bend = int(items[1]) * 10
                        except:
                            pass
                        if len(items) > 2:
                            url = items[1].strip()
                        else: #if info:
                            url = 'javascript://'
                edge = [name_from_id, name_to_id, bend, strength, url, info]
                edges.append(edge)
            else:
                n_pos = line.find('(')
                if n_pos > 0:
                    name = line[:n_pos].strip()
                    com = re.search(r'\(.*\)', line)
                    if com:
                        raw_params = line[com.start()+1:com.end()-1].split(';')
                        params = []
                        params.append(int(raw_params[0]))
                        params.append(int(raw_params[1]))
                        for p in raw_params[2:]:
                            params.append(p.strip())
                        if len(raw_params) < 3:
                            params.append('')
                            params.append(base_color)
                            params.append('')
                        elif len(raw_params) < 4:
                            params.append(base_color)
                            params.append('')
                        elif len(raw_params) < 5:
                            params.append(base_color)

                    if not params[4] or params[4] == '#':
                        params[4] = 'javascript://'
                    d = (R - 10)*2
                    node = [len(nodes), d*params[0], d*params[1], params[3], params[4], name, params[2]]
                nodes.append(node)

    # HTML template

    data = f"""
    <div class="row mt-3">
        <div class="col-md-9">
            <svg id="svg_{UID}" width="600px" height="400px" viewbox="0 0 600 400"></svg>
        </div>
        <div class="col-md-3">
            <div id="svg_{UID}_info"></div>
        </div>
    </div>
    <script>
        var rcld_{UID} = new RoughCld("svg_{UID}")
        nodes = {nodes}
        edges = {edges}
        rcld_{UID}.draw_cld(nodes, edges, true)
    </script>

    ----------------------------------------------------------
        
    <div class="row mt-3">
        <div class="col-md-9">
            <svg id="svg_{UID}" width="800px" height="600px"></svg>
        </div>
        <div class="col-md-3">
            <div id="svg_{UID}_info"></div>
        </div>
    </div>
    <div class="col-12">
        <div class="btn-group" role="group" aria-label="forward / backward">
            <button class="btn btn-light" onclick="rcld_{UID}.prev_edge()">
                <span class="fas fa-chevron-circle-left"></span>
            </button>
            <button class="btn btn-light" onclick="rcld_{UID}.next_edge()">
                <span class="fas fa-chevron-circle-right"></span>
            </button>
        </div>
    </div>
    <script>
        var rcld_{UID} = new RoughCld("svg_{UID}")
        nodes = {nodes}
        edges = {edges}
        rcld_{UID}.draw_cld(nodes, edges, true)
    </script>
"""
    return data.replace("'", '"')


# read input file and generate output file
if __name__ == "__main__":
    if len(sys.argv) < 2:
        print ('Usage: python generate_cld.py <filename>')
    else:
        filename = sys.argv[1]
        n_pos = filename.rfind('.')
        if n_pos > 0:
            f_name = filename[:n_pos]
        else:
            f_name = filename
        f_name += '.html'

        with open(filename, 'r') as f:
            causal_raw = f.read()
        causal_text = generate_causal_diagram(causal_raw)
        with open(f_name, 'w') as f:
            f.write(causal_text)
        print (f'Result: {f_name}')
